src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs

```diff
@@ -815,40 +817,107 @@ internal static int GetHashCodeOrdinalIgnoreCase(ReadOnlySpan<char> value)
         // or are otherwise mitigated
         internal unsafe int GetNonRandomizedHashCode()
         {
+            uint hash1 = (5381 << 16) + 5381;
+            uint hash2 = hash1;
+            uint hash3 = hash1;
+            uint hash4 = hash1;
+
             fixed (char* src = &_firstChar)
             {
                 Debug.Assert(src[this.Length] == '\0', "src[this.Length] == '\\0'");
-                Debug.Assert(((int)src) % 4 == 0, "Managed string should start at 4 bytes boundary");
-
-                uint hash1 = (5381 << 16) + 5381;
-                uint hash2 = hash1;
+                Debug.Assert(((int) src) % 4 == 0, "Managed string should start at 4 bytes boundary");
 
-                uint* ptr = (uint*)src;
+                uint* ptr = (uint*) src;
                 int length = this.Length;
 
-                while (length > 2)
+                if (Vector128.IsHardwareAccelerated && length >= 2 * Vector128<ushort>.Count)
+                {
+                    Vector128<uint> hashVector = Vector128.Create(hash1);
+
+                    while (length > 8)
+                    {
+                        Vector128<uint> srcVec = Vector128.Load(ptr);
+                        length -= 8;
+                        hashVector = (hashVector + RotateLeft(hashVector, 5)) ^ srcVec;
+                        ptr += 4;
+                    }
+
+                    uint hashed1 = hashVector.GetElement(0);
+                    uint hashed2 = hashVector.GetElement(1);
+                    uint hashed3 = hashVector.GetElement(2);
+                    uint hashed4 = hashVector.GetElement(3);
+
+                    while (length > 4)
```

Comment:
I am not sure why you have changes this to `while` loop. I think `if` was perfectly fine here and more efficient too.

Reply:
The original main loop processes the data stream in the way:
```
while(length > 2)
{
   length -= 4
}
```

There will be an extra iteration when length is 4n-3, e.g. 7. consuming an extra null terminator, and it also leaves the fact that the trailing string length can only 0/1/2, so it can be processed by the following `if` statement.

While in the updated case, since the main loop is operating on a larger granularity, the same trick might violate the memory beyond the null terminator. So I used `while` here.

Reply:
Can the condition in the main while loop be `while (length >= 8)` and the check `if (length >= 4)`?

I understand that the existing code does tricks with the null terminator to save a few instructions. You do not have to match those tricks.

Reply:
Sure, thanks for the suggestion, I can try that.